docs/madr/decisions/017-mesh-proxy-template.md

```diff
@@ -0,0 +1,244 @@
+
+This is not required to deliver a new policy, but it would improve the UX. We can implement this as an improvement.
+
+### Routes are hard to modify
+
+We allow to modify VirtualHost, in which you can replace list of routes. You cannot modify a single route.
+You also cannot modify routes if routes are delivered through RDS.
+
+### Merging of TypedConfig
+
+Some parts in Envoy Config are encoded as TypedConfig (marshaled proto in proto).
+Merging such configs results in just replacing. That's why we cannot currently patch TransportSocketMatch.
+
+### Appending to lists
+
+The default strategy for list is replace. Sometimes we may want to add something to the list. There is no such option right now.
+
+### Other resources
+
+We could support TransportSocketMatch https://github.com/kumahq/kuma/issues/4948
+
+## Considered Options
+
+* MeshProxyTemplate
+
+## Decision Outcome
+
+Chosen option: "MeshProxyTemplate".
+
+### MeshProxyTemplate
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: MeshProxyTemplate
+metadata:
+  name: custom-template-1
+  namespace: kuma-system
+  labels:
+    kuma.io/mesh: default
+spec:
+  targetRef:
+    - kind: Mesh
+  defaults:
+    appendModifications:
+      - cluster:
+          operation: add
+          value: |
+            name: test-cluster
+            connectTimeout: 5s
+            type: STATIC
+      - cluster:
+          operation: patch
+          match: # optional: if absent, all clusters will be patched
+            name: test-cluster # optional: if absent, all clusters regardless of name will be patched
+            origin: inbound # optional: if absent, all clusters regardless of its origin will be patched
+          value: | # you can specify only part of cluster definition that will be merged into existing cluster
+            connectTimeout: 5s
+          jsonPatches: # either jsonPatches or value. Only possible to use with operation: patch. See https://jsonpatch.com/ for more.
+            - op: add
+              from: ""
+              path: "/connectTimeout"
+              value: "5s"
+      - cluster:
+          operation: remove
+          match: # optional: if absent, all clusters will be removed
+            name: test-cluster # optional: if absent, all clusters regardless of name will be removed
+            origin: inbound # optional: if absent, all clusters regardless of its origin will be removed
+```
+
+A couple of things to note when you compare this to the old policy:
+* `imports` is gone, solving _Profile imports section_ problem.
+* `appendModifications` allows to append modifications. This solves _No aggregation_ problem.
+* Because list are appended, there is no way to override or exclude modifications.
+  This should be fine for this time being, since proxy templates are managed by the mesh operator.
+  We could introduce `modifications` next to `appendModifications` or provide more general mechanism for excluding/overriding.
```

Comment:
Can't you always introduce a new modification that undoes (via operations/JSON patch) an earlier one?

Reply:
Yes, you can. Although it's more complicated than just wiping existing modifications